File: Frontend-admin/src/types/producto.ts
import type { Ingrediente } from './ingrediente'
import type { Categoria } from './categoria'
import type { UnidadMedida } from './unidad_medida'

export interface ProductoIngredienteLink {
  ingrediente_id: number
  cantidad: number
  unidad_medida_id: number
  es_removible: boolean
  ingrediente?: Ingrediente
}

export interface ProductoCategoriaLink {
  categoria_id: number
  es_principal: boolean
  categoria?: Categoria
}

export interface Producto {
  id: number
  nombre: string
  descripcion: string | null
  precio_base: number
  imagenes_url: string | null
  stock_cantidad: number
  disponible: boolean
  unidad_venta_id: number | null
  unidad_venta?: UnidadMedida
}

export interface ProductoDetalle extends Producto {
  ingrediente_links: ProductoIngredienteLink[]
  categoria_links: ProductoCategoriaLink[]
}

export interface ProductoCategoriaLinkCreate {
  categoria_id: number
  es_principal: boolean
}

export interface ProductoIngredienteLinkCreate {
  ingrediente_id: number
  cantidad: number
  unidad_medida_id: number
  es_removible: boolean
}

export interface ProductoCreate {
  nombre: string
  descripcion?: string
  precio_base: number
  imagenes_url?: string
  stock_cantidad?: number
  disponible?: boolean
  unidad_venta_id?: number
}

export interface ProductoUpdate {
  nombre?: string
  descripcion?: string
  precio_base?: number
  imagenes_url?: string
  stock_cantidad?: number
  disponible?: boolean
  unidad_venta_id?: number
}

File: Backend/app/schemas/producto.py
from pydantic import BaseModel, Field
from typing import Optional

from .ingrediente import IngredienteRead
from .categoria import CategoriaRead
from .unidad_medida import UnidadMedidaRead


class ProductoIngredienteLinkRead(BaseModel):
    ingrediente_id: int
    cantidad: float
    unidad_medida_id: int
    es_removible: bool
    ingrediente: Optional[IngredienteRead] = None

    model_config = {"from_attributes": True}


class ProductoCategoriaLinkRead(BaseModel):
    categoria_id: int
    es_principal: bool
    categoria: Optional[CategoriaRead] = None

    model_config = {"from_attributes": True}


class ProductoCategoriaLinkCreate(BaseModel):
    categoria_id: int
    es_principal: bool = False


class ProductoIngredienteLinkCreate(BaseModel):
    ingrediente_id: int
    cantidad: float = Field(gt=0)
    unidad_medida_id: int
    es_removible: bool = False


class ProductoCreate(BaseModel):
    nombre: str = Field(min_length=1, max_length=150)
    descripcion: Optional[str] = None
    precio_base: float = Field(ge=0)
    imagenes_url: Optional[str] = None
    stock_cantidad: int = Field(default=0, ge=0)
    disponible: bool = True
    unidad_venta_id: Optional[int] = None


class ProductoUpdate(BaseModel):
    nombre: Optional[str] = Field(default=None, min_length=1, max_length=150)
    descripcion: Optional[str] = None
    precio_base: Optional[float] = Field(default=None, ge=0)
    imagenes_url: Optional[str] = None
    stock_cantidad: Optional[int] = Field(default=None, ge=0)
    disponible: Optional[bool] = None
    unidad_venta_id: Optional[int] = None


class ProductoRead(BaseModel):
    id: int
    nombre: str
    descripcion: Optional[str]
    precio_base: float
    imagenes_url: Optional[str]
    stock_cantidad: int
    disponible: bool
    unidad_venta_id: Optional[int]
    unidad_venta: Optional[UnidadMedidaRead] = None

    model_config = {"from_attributes": True}


class ProductoReadDetalle(ProductoRead):
    ingrediente_links: list[ProductoIngredienteLinkRead] = []
    categoria_links: list[ProductoCategoriaLinkRead] = []
